sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +301,183 @@ def reduce(self, word, exclude=None):
+
+        Returns:
+            Dictionary of inverse_rules.
+
+        '''
+        inverse_rules = {}
+        for r in rules:
+            rule_key_inverse = r**-1
+            rule_value_inverse = (rules[r])**-1
+            if (rule_value_inverse < rule_key_inverse):
+                inverse_rules[rule_key_inverse] = rule_value_inverse
+            else:
+                inverse_rules[rule_value_inverse] = rule_key_inverse
+        return inverse_rules
+
+    def construct_automaton(self):
+        '''
+        Construct the automaton based on the set of reduction rules of the system.
+
+        Automata Design:
+        The accept states of the automaton are the proper prefixes of the left hand side of the rules.
+        The complete left hand side of the rules are the dead states of the automaton.
+
+        '''
+        self._add_to_automaton(self.rules)
+
+    def _add_to_automaton(self, rules):
+        '''
+        Add new states and transitions to the automaton.
+
+        Summary:
+        States corresponding to the new rules added to the system are computed and added to the automaton.
+        Transitions in the previously added states are also modified if necessary.
+
+        Arguments:
+            rules (dictionary) -- Dictionary of the newly added rules.
+
+        '''
+        # Automaton variables
+        automaton_alphabet = []
+        proper_prefixes = {}
+
+        # compute the inverses of all the new rules added
+        all_rules = rules
+        inverse_rules = self._compute_inverse_rules(all_rules)
+        all_rules.update(inverse_rules)
+
+        # Keep track of the accept_states.
+        accept_states = []
+
+        # The symbols present in the new rules are the symbols to be verified at each state.
+        # computes the automaton_alphabet, as the transitions solely depend upon the new states.
+        for rule in all_rules:
+            automaton_alphabet += rule.letter_form_elm
+        automaton_alphabet = set(automaton_alphabet)
+
+        # Compute the proper prefixes for every rule.
+        for r in all_rules:
+            proper_prefixes[r] = []
+            letter_word_array = [s for s in r.letter_form_elm]
+            for i in range (1, len(letter_word_array)):
+                letter_word_array[i] = letter_word_array[i-1]*letter_word_array[i]
+            proper_prefixes[r] = letter_word_array
+
+        # Add dead states.
+        # This is essentially checking the left hand side of the rules.
+        for rule in all_rules:
+            if not rule in self.reduction_automaton.states:
+                self.reduction_automaton.add_state(rule, state_type='d', rh_rule=all_rules[rule])
+
+        for r in all_rules:
+            prop_prefix = proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in self.reduction_automaton.states:
+                    self.reduction_automaton.add_state(elem, state_type='a')
+                    accept_states.append(elem)
```

Comment:
(Okay, I know I said that there shouldn't be any more changes to make but now that I've noticed multiple small inefficiencies in this part of the code, I can't just let them pass. I promise what I suggest here will be the absolutely last changes and this will be good to merge)

The last 4 for loops all iterate over `all_rules` - it can all be done in one loop, and you won't need to store a `proper_prefixes` in that loop because you can insert what follows after `for elem in prop_prefix` in the part where you would compute the prefixes. Do you see this?